Storage/SQL/SQLiteHistoryRecommendations.swift

```diff
@@ -50,60 +50,30 @@ extension SQLiteHistory: HistoryRecommendations {
         return self.db.runQuery(sql, args: nil, factory: SQLiteHistory.iconHistoryMetadataColumnFactory)
     }
 
-    public func invalidateHighlights() -> Success {
-        return clearHighlights() >>> populateHighlights
-    }
-
     public func removeHighlightForURL(_ url: String) -> Success {
         return self.db.run([("INSERT INTO \(TableActivityStreamBlocklist) (url) VALUES (?)", [url])])
     }
 
-    public func clearHighlights() -> Success {
-        return self.db.run("DELETE FROM \(TableHighlights)", withArgs: nil)
+    public func repopulateHighlights() -> Success {
+        return self.db.run(self.repopulateHighlightsQuery())
     }
 
-    private func populateHighlights() -> Success {
+    private func repopulateHighlightsQuery() -> [(String, Args?)] {
         let (query, args) = computeHighlightsQuery()
+        let clearHighlightsQuery = "DELETE FROM \(TableHighlights)"
 
-        // Convert the fetched row into arguments for a bulk insert along with the
-        // generated cache_key value.
-        func argsFrom(row: SDRow) -> Args? {
-            let urlString = row["url"] as! String
-            let cacheKey = SQLiteMetadata.cacheKeyForURL(urlString.asURL!)!
-            return [
-                row["historyID"],
-                cacheKey,
-                urlString,
-                row["title"],
-                row["guid"],
-                row["visitCount"],
-                row["visitDate"],
-                row["is_bookmarked"]
-            ]
-        }
-        
-        // Run the highlights computation query and take the results to bulk insert into the cached highlights table
-        return self.db.runQuery(query, args: args, factory: argsFrom)
-            >>== { highlightRows in
-                let values: [Args] = highlightRows.asArray().flatMap { $0 }
-                let highlightsProjection = [
-                    "historyID",
-                    "cache_key",
-                    "url",
-                    "title",
-                    "guid",
-                    "visitCount",
-                    "visitDate",
-                    "is_bookmarked"
-                ]
-
-                return self.db.bulkInsert(
-                    TableHighlights,
-                    op: .InsertOrReplace,
-                    columns: highlightsProjection,
-                    values: values
-            )
+        let sql = "INSERT INTO \(TableHighlights) " +
```

Comment:
You might want to explicitly specify `(historyID, cache_key, url…)` here so that if the schema changes this insert will fail rather than silently doing the wrong thing.